client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1035 @@
+    }
+    if (!new_subscription) {
+        KAA_FREE(subs_topic_id);
+        return KAA_ERR_NOMEM;
+    }
+    self->subscriptions = new_subscription;
+    if (force_sync) {
+        KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Going to subscribe to topic %lu", *topic_id);
+        kaa_sync_topic_subscriptions(self);
+    } else {
+        KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Subscription to topic %lu is postponed till sync", *topic_id);
+    }
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_subscribe_to_topics(kaa_notification_manager_t *self, uint64_t *topic_ids, size_t size, bool force_sync)
+{
+    KAA_RETURN_IF_NIL2(self, topic_ids, KAA_ERR_BADPARAM);
+    kaa_list_t *new_subscription = NULL;
+    kaa_topic_t *topic = NULL;
+    while (size--) {
+        kaa_error_t err = kaa_find_topic(self, &topic, topic_ids + size);
+        if (err) {
+            KAA_LOG_WARN(self->logger, KAA_ERR_BADPARAM, "Failed to subscribe to the topic with id = %lu.", topic_ids[size]);
+            return err;
+        } else {
+            if (topic->subscription_type == MANDATORY) {
+                KAA_LOG_WARN(self->logger, KAA_ERR_BADPARAM, "Failed to subscribe to the topic with id = %lu. Topic isn't optional.", topic_ids[size]);
+                return KAA_ERR_BADPARAM;
+            }
+            uint64_t *subs_topic_id = (uint64_t *) KAA_MALLOC(sizeof(uint64_t));
+            KAA_RETURN_IF_NIL(subs_topic_id, KAA_ERR_NOMEM);
+            *subs_topic_id = topic_ids[size];
+            if (!self->subscriptions) {
+                new_subscription = kaa_list_create(subs_topic_id);
+            } else {
+                new_subscription = kaa_list_push_front(self->subscriptions, subs_topic_id);
+            }
+            if (!new_subscription) {
+                KAA_FREE(subs_topic_id);
+                return KAA_ERR_NOMEM;
+            }
+            self->subscriptions = new_subscription;
+        }
+    }
+    if (force_sync) {
+        KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Going to subscribe to the topics");
+        kaa_sync_topic_subscriptions(self);
+    } else {
+        KAA_LOG_INFO(self->logger, KAA_ERR_NONE, "Subscription to topics is postponed till sync");
+    }
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_unsubscribe_from_topic(kaa_notification_manager_t *self, uint64_t *topic_id, bool force_sync)
+{
+    KAA_RETURN_IF_NIL2(self, topic_id, KAA_ERR_BADPARAM);
+    kaa_topic_t* topic = NULL;
+    kaa_error_t err = kaa_find_topic(self, &topic, topic_id);
+    if (err) {
+        KAA_LOG_WARN(self->logger, KAA_ERR_BADPARAM, "Failed to unsubscribe from the topic with id = %lu.", *topic_id);
+        return err;
+    }
+    if (topic->subscription_type == MANDATORY) {
+        KAA_LOG_WARN(self->logger, KAA_ERR_BADPARAM, "Failed to unsubscribe from the topic with id = %lu. Topic isn't optional.", *topic_id);
+        return KAA_ERR_BADPARAM;
+    }
+    uint64_t *unsubs_topic_id = (uint64_t *) KAA_MALLOC(sizeof(uint64_t));
```

Comment:
Almost duplicate code block. See kaa_subscribe_from_topic function. Please, try to reduce code amount.